### ISSUE TITLE
Improve unnecessary translation loading

### ISSUE THREAD

**Issue by aldeed** [reactioncommerce/reaction#3284]
Something I noticed while working on SimpleSchema changes...

`loadTranslations` runs for each plugin with translations, every time the server restarts, doing a ton of upserts that are essentially no-ops. This *might* not be necessary on every startup in production, but it's *definitely* not necessary on every startup while developing.

The solution would probably involve some combination of caching/comparing a checksum and/or testing whether we're in production mode. Might also help to have a version in each json file.

Or the whole thing could be triggered manually by an admin instead of on startup.

I haven't given it enough thought to know what's best, but I wanted to mention it before I forget.

Potentially dynamic imports solves this issue.

Consider adding a configuration file for enabled languages. 

Import languages once a shop has toggled on i18n.

Hook `i18n/flushTranslations` up to a button in reaction devtools to make resetting translations easily.

**Comment by zenweasel** [comment]
@aldeed Should this get the `performance` label?

**Comment by mikemurray** [comment]
Findings

Translations are loaded into the Assets collection using a bulk operation on the raw Mongo collection. They are subsequently loaded into the Translations collection using the Reaction `Import` which is another bulk importer.

There seems to be a deviation of hundreds of milliseconds between multiple soft reloads with the only change being an updated `console.log` to trigger a build.

Having the translation import enabled adds 2.5 to 3 seconds to the reload.

With multiple shops, while it does take longer to start the app, the translations import is still 2.5 to 3 seconds added on top of the longer load.

### Profiling

```
// One Shop - soft reloads
// Off
other Meteor.startup at main.js (server/main.js:12:8)....1,263 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....1,189 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....1,312 ms (1)

// On
other Meteor.startup at main.js (server/main.js:12:8)....3,904 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....3,964 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....4,046 ms (1)
```
---

**begin transition to 6 shops**

```
// Commented out code to disable translations - soft reload
other Meteor.startup at main.js (server/main.js:12:8)...11,689 ms (1)

// Fresh restart, killed meteor and restarted with translations off
other Meteor.startup at main.js (server/main.js:12:8)....9,209 ms (1) 
```

---
```
// Five Merchant shops + Main shop - soft reloads
// Off
other Meteor.startup at main.js (server/main.js:12:8)....5,782 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....5,928 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....6,028 ms (1)

// On
other Meteor.startup at main.js (server/main.js:12:8)....8,565 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....8,533 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....8,639 ms (1)
```

**Conclusion**

- Translation load adds 3 seconds or less to startup for soft resets for 1 or many shops.
- App load/reload times rise as more shops are enabled.
- App reload times to be affected by massive code changes (commented out)

**Comment by aldeed** [comment]
Interesting findings. Two things to note:

1. The reason for taking this out of startup is not just due to the time it takes, but also due to unnecessary DB ops, which = $$ and potential slowness on the DB side.
2. If you commented out the DB upserts but not the i18n imports, you might not be capturing the full effect. If comment out all the `import "./i18n";` lines like in /imports/plugins/core/i18n/server/index.js, this may make the Babel compile step a lot faster. And then those files in /i18n folders can be loaded only when necessary using `fs` utils instead.

**Comment by mikemurray** [comment]
To use FS to load these i18n files, they will have to be included in the final bundle, for production. As far as I know, with the imports directory, you'll have to have imported the files you want to be included with the production bundle from a file outside the imports directory, like the `plugins.js` files in `client` and `server`.

In development mode it's fine since we have access to the filesystem, but once everything it built, we won't have access to those translation files. Some translations still live in "/private" and we do use FS to load them.

Removing the actual imports could help the Babel transilpation run faster, and where would be 1000 less files to watch by Meteor.

---

Reducing DB ops could be as simple checking if translation have been loaded into the Assets collection, and if they are, then stop there. Then theres a button to flush and reload those files Assets and let the import run as usual.

I cannot speak to the actual cost or performance of those DB operations.

@spencern thoughts?

**Comment by spencern** [comment]
Let's check to see what kind of a difference it makes the compile time if the i18n files are commented out.

**Comment by mikemurray** [comment]
Doing so now.

**Comment by mikemurray** [comment]
# Disabled all `import "./i18n";`
## Run 1
### fresh startup from 1.8.0

```
Build App...................................................161,719 ms (1)
Babel.compile...........................................79,772 ms (1587)
Server startup...............................................19,722 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....4,419 ms (1)
```

### soft-reload after code change

```
Rebuild App..................................................33,933 ms (1)
Babel.compile..............................................411 ms (1587)
Server startup................................................5,433 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....1,371 ms (1)
```


## Run 2
### fresh startup from 1.8.0

```
Build App...................................................159,624 ms (1)
Babel.compile...........................................79,786 ms (1587)
Server startup...............................................18,714 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....4,182 ms (1)
```

### soft-reload after code change

```
Rebuild App..................................................33,791 ms (1)
Server startup................................................6,147 ms (1)
Babel.compile..............................................310 ms (1587)
other Meteor.startup at main.js (server/main.js:12:8)....1,568 ms (1)
```


---


# Enabled all `import "./i18n";`
## Run 1
### fresh startup from 1.8.0

```
Build App...................................................174,787 ms (1)
Babel.compile...........................................80,050 ms (1587)
Server startup...............................................21,706 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....6,082 ms (1)
```

### soft-reload after code change

```
Rebuild App..................................................34,958 ms (1)
Babel.compile..............................................308 ms (1587)
Server startup................................................8,501 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....3,873 ms (1)
```


## Run 2
### fresh startup from 1.8.0

```
Build App...................................................201,649 ms (1)
Babel.compile...........................................78,900 ms (1587)
Server startup...............................................21,724 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....6,247 ms (1)
```

### soft-reload after code change

```
Rebuild App..................................................47,883 ms (1)
Babel.compile..............................................255 ms (1587)
Server startup................................................8,767 ms (1)
other Meteor.startup at main.js (server/main.js:12:8)....3,475 ms (1)
```

**Comment by mikemurray** [comment]
- Startup is slow regardless of imports, just slightly faster with import files not imported at all in JS
- Build/rebuild times for babel are similar
- App rebuild times are slightly slower with imports included, but not by much
- Fluctuations between Run 1 and Run 2 of imports enabled is confusion. Perhaps over time, it degrades?


Sample size to very small, but hopefully it helps illustrate the performance of having the i18n imports actually imported in JS vs not.

**Comment by aldeed** [comment]
@mikemurray That's less improvement than I had hoped. Do you think Meteor might be watching them just by their existing? I'm pretty sure it's not supposed to do that within /imports, but it would be interesting to see if there's any improvement with all the i18n json files deleted.

You're right that they'd have to move to be imported by fs. Dynamic import is also a good solution if we deem it's worth doing.

**Comment by spencern** [comment]
Closed by #3838
